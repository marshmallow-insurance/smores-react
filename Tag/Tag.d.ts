import { FC } from 'react';
import { MarginProps } from '../utils/space';
import { Color } from '../theme';
export type TagProps = {
    label: string;
    color: Color;
    bgColor: Color;
    borderColor?: Color;
    className?: string;
} & MarginProps;
export declare const Tag: FC<TagProps>;
