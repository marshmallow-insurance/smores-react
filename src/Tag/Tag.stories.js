import React from 'react'
import { storiesOf } from '@storybook/react'
import { text } from '@storybook/addon-knobs'

import { Tag } from './Tag'
import { theme } from '../theme'

storiesOf('Tag', module).addWithJSX('default', () => (
  <Tag
    label={text('Label', 'Covered')}
    bgColor={theme.colors.green[500]}
    borderColor={theme.colors.green[700]}
    color={text('Color', 'white')}
    className={text('Class name', '')}
  />
))
