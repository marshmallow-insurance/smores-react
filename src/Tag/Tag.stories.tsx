import React from 'react'
import { Tag, TagProps } from './Tag'
import { CollectionPage } from './Collection'

export default {
  title: 'Tag',
  component: Tag,
}

const Template = (props: TagProps) => <Tag {...props} />

export const Default = Template.bind({})

Default.args = {
  label: 'This is a tag',
  bgColor: 'feijoa',
  borderColor: 'feijoa',
  color: 'cream',
}

Default.argTypes = {
  theme: { table: { disable: true } },
  as: { table: { disable: true } },
  forwardedAs: { table: { disable: true } },
}

export const Collection = CollectionPage.bind({})
