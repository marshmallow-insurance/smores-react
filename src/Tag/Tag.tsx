import React, { FC } from 'react'
import styled, { useTheme } from 'styled-components'
import { MarginProps } from '../utils/space'

import { TransientProps } from 'utils/utilTypes'
import { Box } from '../Box'
import { Text } from '../Text'
import { Icon } from '../Icon'
import { Icons } from 'Icon/iconsList'
import {
  ColorTypes,
  getColorPath,
  resolveToThemeColor,
} from '../ThemeProvider/utils/colourMap'

export type TagProps = {
  label: string
  color: ColorTypes
  bgColor: ColorTypes
  borderColor?: ColorTypes
  className?: string
  icon?: Icons
  iconColor?: ColorTypes
  onClick?: () => void
} & MarginProps

export const Tag: FC<TagProps> = ({
  label,
  color,
  borderColor,
  bgColor,
  className,
  icon,
  iconColor,
  onClick,
  ...marginProps
}) => {
  const theme = useTheme()

  const resolvedBGColor = resolveToThemeColor(bgColor, theme)
  const resolvedBorderColor =
    borderColor && resolveToThemeColor(borderColor, theme)

  return (
    <Wrapper
      $bgColor={resolvedBGColor}
      className={className}
      $borderColor={resolvedBorderColor}
      {...marginProps}
      alignContent="center"
      justifyContent="center"
      onClick={onClick}
    >
      {icon && (
        <TagIcon
          render={icon}
          color={iconColor}
          size={16}
          data-testid={`tag-icon-${icon}`}
        />
      )}
      <TagText tag="span" typo="label" color={getColorPath(color)}>
        {label}
      </TagText>
    </Wrapper>
  )
}

type WrapperProps = TransientProps<{
  bgColor: string
  borderColor?: string
}>

const Wrapper = styled(Box)<WrapperProps>`
  background-color: ${({ $bgColor }) => $bgColor};
  border: ${({ $borderColor }) => $borderColor && `1px solid ${$borderColor}`};

  border-radius: 6px;
  padding: 4px 8px;

  height: 22px;

  box-sizing: border-box;
  display: inline-flex;
`

const TagText = styled(Text)`
  display: flex;
  align-items: center;
  font-size: 10px;
  font-weight: 500;
  line-height: 12px;
  text-transform: uppercase;
`

const TagIcon = styled(Icon)`
  padding-right: 4px;
`
