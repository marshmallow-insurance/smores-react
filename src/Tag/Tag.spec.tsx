import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Tag } from '../Tag'
import { TagProps } from './Tag'

describe('Tag Component', () => {
  testCases.forEach(({ description, props }) => {
    test(description, () => {
      render(<Tag {...props} />)

      expect(screen.getByText(props.label)).toBeInTheDocument()

      if (props.icon) {
        const iconElement = screen.getByTestId(`tag-icon-${props.icon}`)
        expect(iconElement).toBeInTheDocument()
      } else {
        const iconElement = screen.queryByTestId(`tag-icon-${props.icon}`)
        expect(iconElement).not.toBeInTheDocument()
      }
    })
  })
})

const testCases = [
  {
    description: 'Renders Tag with label Test Tag 0001',
    props: {
      label: 'Test Tag 0001',
      color: 'apple',
      bgColor: 'chia',
    } as TagProps,
  },
  {
    description: 'Renders Tag with label Test Tag 0002 and flag icon',
    props: {
      label: 'Test Tag 0002',
      color: 'apple',
      bgColor: 'chia',
      icon: 'flag',
      iconColor: 'blueberry',
    } as TagProps,
  },
  {
    description:
      'Renders Tag with label Test Tag 0003 and no icon if iconColor: blueberry, icon: undefined',
    props: {
      label: 'Test Tag 0003',
      color: 'apple',
      bgColor: 'chia',
      iconColor: 'blueberry',
    } as TagProps,
  },
  {
    description:
      'Renders Tag with label Test Tag 0004 and icon with default color if iconColor: undefined',
    props: {
      label: 'Test Tag 0004',
      color: 'apple',
      bgColor: 'chia',
      icon: 'flagFilled',
    } as TagProps,
  },
]
