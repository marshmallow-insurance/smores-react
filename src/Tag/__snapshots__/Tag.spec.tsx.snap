// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tag Component > renders snapshot test correctly 1`] = `
<div>
  <div
    class="sc-bRKDuR gQglmM sc-jwTyAe iDwEeu"
  >
    <span
      class="sc-bRKDuR bHjpxP sc-dTvVRJ cWoRvt sc-hjsuWn juqWeu"
      cursor="inherit"
      title=""
    >
      Snapshot Test Tag
    </span>
  </div>
</div>
`;
